exercises/t_jubb/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,270 @@
+		perimeter = M_PI * 2 * radius;
+	};
+
+private:
+	double radius;
+};
+
+/*
+Comparisons between shapes for area, perimeter
+*/
+bool compare_area(Shape& shape1, Shape& shape2) {
+	return shape1.get_area() < shape2.get_area();
+};
+bool compare_perimeter(Shape& shape1, Shape& shape2) {
+	return shape1.get_perimeter() < shape2.get_perimeter();
+};
+
+class ShapeSorter : public Shape {
+
+public:
+	ShapeSorter() { };
+	ShapeSorter(std::vector<Shape> vec) { shape_vector = vec; };
+	void print_type(std::string type);
+	void print_sides(int sides);
+	void print_by_area(bool ascending);
+	void print_by_perimeter(bool ascending);
+
+private:
+	std::vector<Shape> shape_vector;
+
+};
+
+/*
+Prints shapes in shape_vector with a given type.
+*/
+void ShapeSorter::print_type(std::string type) {
+	for (auto i = shape_vector.begin(); i != shape_vector.end(); i++) {
+		if ((*i).get_type() == type) {
+			(*i).print_shape();
+		};
+	};
+};
+
+/*
+Prints shapes in shape_vector with a given number of sides.
+*/
+void ShapeSorter::print_sides(int sides) {
+	for (auto i = shape_vector.begin(); i != shape_vector.end(); i++) {
+		if ((*i).get_sides() == sides) {
+			(*i).print_shape();
+		};
+	};
+};
+
+/*
+
+Sort the vector of Shapes by their area and print out the contents.
+
+@param ascending : Wether to sort from lowest to highst
+
+*/
+void ShapeSorter::print_by_area(bool ascending) {
+
+	sort(shape_vector.begin(), shape_vector.end(),compare_area);
+
+	if (ascending) {
+		for (auto i = shape_vector.begin(); i != shape_vector.end(); i++) {
+			(*i).print_shape();
+		};
+	}
+	else {
+		for (auto i = shape_vector.rbegin(); i != shape_vector.rend(); i++) {
+			(*i).print_shape();
+		};
+
+	}
+	
+};
+
+/*
+
+Sort the vector of Shapes by their perimeter and print out the contents.
+
+@param ascending : Wether to sort from lowest to highst
+
+*/
+void ShapeSorter::print_by_perimeter(bool ascending) {
+
+	sort(shape_vector.begin(), shape_vector.end(), compare_perimeter);
+
+	if (ascending) {
+		for (auto i = shape_vector.begin(); i != shape_vector.end(); i++) {
+			(*i).print_shape();
+		};
+	}
+	else {
+		for (auto i = shape_vector.rbegin(); i != shape_vector.rend(); i++) {
+			(*i).print_shape();
+		};
+
+	}
+
+};
+
+
+int main(int, char **)
+{
+	std::cout << "Hello World" << std::endl;
+
+	Shape shape;
+	std::cout << shape.get_area() << std::endl;
+	std::cout << shape.get_perimeter() << std::endl;
+	std::cout << shape.get_sides() << std::endl;
+
+	Circle ci(5.0);
+	std::cout << "Area      : " << ci.get_area() << std::endl;
+	std::cout << "Perimeter : " << ci.get_perimeter() << std::endl;
+	std::cout << "Sides     : " << ci.get_sides() << std::endl;
+
+	Square sq(1.0);
+	std::cout << "Area      : " << sq.get_area() << std::endl;
+	std::cout << "Perimeter : " << sq.get_perimeter() << std::endl;
+	std::cout << "Sides     : " << sq.get_sides() << std::endl;
+
+	Rectangle re(1.0, 2.0);
+	IsoscelesTriangle tri(5.7, 2.3);
+
+	std::vector<Shape> vec;
```

Comment:
The way your code is currently structured this does work but if you change it to the abstract implementation discussed above then this won't work as expected due to something called [object slicing](https://en.wikipedia.org/wiki/Object_slicing). This is related to [polymorphism](http://www.cplusplus.com/doc/tutorial/polymorphism/) and how you can store vectors of things that are related by a similar concept; in this case Shape.